File: Core/src/main/java/com/hqq/core/widget/divider/DividerBuilder.java
package com.hqq.core.widget.divider;

import android.content.Context;
import android.support.annotation.ColorInt;
import android.support.v4.content.ContextCompat;

import com.hqq.core.R;

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.widget.divider
 * @FileName :   DividerBuilder
 * @Date : 2018/7/5 0005  下午 5:16
 * @Descrive : TODO
 * @Email :
 */
public class DividerBuilder {
    private SideLine leftSideLine;
    private SideLine topSideLine;
    private SideLine rightSideLine;
    private SideLine bottomSideLine;


    public DividerBuilder setLeftSideLine(boolean isHave, @ColorInt int color, float widthDp, float startPaddingDp, float endPaddingDp) {
        this.leftSideLine = new SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp);
        return this;
    }

    public DividerBuilder setLeftSideLine(@ColorInt int color, float widthDp) {
        return setLeftSideLine(true, color, widthDp, 0, 0);
    }

    public DividerBuilder setLeftSideLine(Context context, int color, int widthDp) {
        return setLeftSideLine(true, ContextCompat.getColor(context, color), context.getResources().getDimension(widthDp), 0, 0);
    }

    public DividerBuilder setLeftSideLine(Context context, int widthDp) {
        return setLeftSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.getResources().getDimension(widthDp), 0, 0);
    }


    public DividerBuilder setTopSideLine(boolean isHave, @ColorInt int color, float widthDp, float startPaddingDp, float endPaddingDp) {
        this.topSideLine = new SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp);
        return this;
    }

    public DividerBuilder setTopSideLine(boolean isHave, @ColorInt int color, float widthDp) {
        return setTopSideLine(isHave, color, widthDp, 0, 0);
    }

    public DividerBuilder setTopSideLine(@ColorInt int color, float widthDp) {
        return setTopSideLine(true, color, widthDp, 0, 0);
    }

    public DividerBuilder setTopSideLine(Context context, int color, int widthDp) {
        return setTopSideLine(true, ContextCompat.getColor(context, color), context.getResources().getDimension(widthDp), 0, 0);
    }

    public DividerBuilder setTopSideLine(Context context, int widthDp) {
        return setTopSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.getResources().getDimension(widthDp), 0, 0);
    }


    public DividerBuilder setRightSideLine(boolean isHave, @ColorInt int color, float widthDp, float startPaddingDp, float endPaddingDp) {
        this.rightSideLine = new SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp);
        return this;
    }

    public DividerBuilder setRightSideLine(@ColorInt int color, float widthDp) {
        return setRightSideLine(true, color, widthDp, 0, 0);
    }

    public DividerBuilder setRightSideLine(Context context, int color, int widthDp) {
        return setRightSideLine(true, ContextCompat.getColor(context, color), context.getResources().getDimension(widthDp), 0, 0);
    }

    public DividerBuilder setRightSideLine(Context context, int widthDp) {
        return setRightSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.getResources().getDimension(widthDp), 0, 0);
    }

    public DividerBuilder setBottomSideLine(boolean isHave, @ColorInt int color, float widthDp, float startPaddingDp, float endPaddingDp) {
        this.bottomSideLine = new SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp);
        return this;
    }

    public DividerBuilder setBottomSideLine(@ColorInt int color, float widthDp) {
        return setBottomSideLine(true, color, widthDp, 0, 0);
    }

    public DividerBuilder setBottomSideLine(Context context, int color, int widthDp) {
        return setBottomSideLine(true, ContextCompat.getColor(context, color), context.getResources().getDimension(widthDp), 0, 0);
    }

    public DividerBuilder setBottomSideLine(Context context, int widthDp) {
        return setBottomSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.getResources().getDimension(widthDp), 0, 0);
    }


    public Divider create() {
        //提供一个默认不显示的sideline，防止空指针
        SideLine defaultSideLine = new SideLine(false, 0xff666666, 0, 0, 0);
        leftSideLine = (leftSideLine != null ? leftSideLine : defaultSideLine);
        topSideLine = (topSideLine != null ? topSideLine : defaultSideLine);
        rightSideLine = (rightSideLine != null ? rightSideLine : defaultSideLine);
        bottomSideLine = (bottomSideLine != null ? bottomSideLine : defaultSideLine);
        return new Divider(leftSideLine, topSideLine, rightSideLine, bottomSideLine);
    }

}

File: Core/src/main/java/com/easy/core/widget/divider/DividerBuilder.kt
package com.easy.core.widget.divider

import android.content.Context
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import  com.easy.core.R

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.widget.divider
 * @FileName :   DividerBuilder
 * @Date : 2018/7/5 0005  下午 5:16
 * @Email :  qiqiang213@gmail.com
 * @Descrive :
 */
class DividerBuilder {
    private var leftSideLine: SideLine? = null
    private var topSideLine: SideLine? = null
    private var rightSideLine: SideLine? = null
    private var bottomSideLine: SideLine? = null
    fun setLeftSideLine(isHave: Boolean, @ColorInt color: Int, widthDp: Float, startPaddingDp: Float, endPaddingDp: Float): DividerBuilder {
        leftSideLine = SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp)
        return this
    }

    fun setLeftSideLine(@ColorInt color: Int, widthDp: Float): DividerBuilder {
        return setLeftSideLine(true, color, widthDp, 0f, 0f)
    }

    fun setLeftSideLine(context: Context, color: Int, widthDp: Int): DividerBuilder {
        return setLeftSideLine(true, ContextCompat.getColor(context, color), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun setLeftSideLine(context: Context, widthDp: Int): DividerBuilder {
        return setLeftSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun setTopSideLine(isHave: Boolean, @ColorInt color: Int, widthDp: Float, startPaddingDp: Float, endPaddingDp: Float): DividerBuilder {
        topSideLine = SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp)
        return this
    }

    fun setTopSideLine(isHave: Boolean, @ColorInt color: Int, widthDp: Float): DividerBuilder {
        return setTopSideLine(isHave, color, widthDp, 0f, 0f)
    }

    fun setTopSideLine(@ColorInt color: Int, widthDp: Float): DividerBuilder {
        return setTopSideLine(true, color, widthDp, 0f, 0f)
    }

    fun setTopSideLine(context: Context, color: Int, widthDp: Int): DividerBuilder {
        return setTopSideLine(true, ContextCompat.getColor(context, color), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun setTopSideLine(context: Context, widthDp: Int): DividerBuilder {
        return setTopSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun setRightSideLine(isHave: Boolean, @ColorInt color: Int, widthDp: Float, startPaddingDp: Float, endPaddingDp: Float): DividerBuilder {
        rightSideLine = SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp)
        return this
    }

    fun setRightSideLine(@ColorInt color: Int, widthDp: Float): DividerBuilder {
        return setRightSideLine(true, color, widthDp, 0f, 0f)
    }

    fun setRightSideLine(context: Context, color: Int, widthDp: Int): DividerBuilder {
        return setRightSideLine(true, ContextCompat.getColor(context, color), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun setRightSideLine(context: Context, widthDp: Int): DividerBuilder {
        return setRightSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun setBottomSideLine(isHave: Boolean, @ColorInt color: Int, widthDp: Float, startPaddingDp: Float, endPaddingDp: Float): DividerBuilder {
        bottomSideLine = SideLine(isHave, color, widthDp, startPaddingDp, endPaddingDp)
        return this
    }

    fun setBottomSideLine(@ColorInt color: Int, widthDp: Float): DividerBuilder {
        return setBottomSideLine(true, color, widthDp, 0f, 0f)
    }

    fun setBottomSideLine(context: Context, color: Int, widthDp: Int): DividerBuilder {
        return setBottomSideLine(true, ContextCompat.getColor(context, color), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun setBottomSideLine(context: Context, widthDp: Int): DividerBuilder {
        return setBottomSideLine(true, ContextCompat.getColor(context, R.color.transparent), context.resources.getDimension(widthDp), 0f, 0f)
    }

    fun create(): Divider {
        //提供一个默认不显示的sideline，防止空指针
        val defaultSideLine = SideLine(false, -0x99999a, 0f, 0f, 0f)
        leftSideLine = if (leftSideLine != null) leftSideLine else defaultSideLine
        topSideLine = if (topSideLine != null) topSideLine else defaultSideLine
        rightSideLine = if (rightSideLine != null) rightSideLine else defaultSideLine
        bottomSideLine = if (bottomSideLine != null) bottomSideLine else defaultSideLine
        return Divider(leftSideLine, topSideLine, rightSideLine, bottomSideLine)
    }
}